packages/exam-engine/src/dom-utils.ts

```diff
@@ -1,15 +1,81 @@
 export const NBSP = '\u00A0'
 
-export function parentElements(element: Element): Element[] {
+export function parentElements(element: Element, selector?: Selector): Element[] {
+  const predicate = selector != null ? mkPredicate(element, selector) : undefined
   const result = []
   while (element.parentElement != null) {
     element = element.parentElement
-    result.push(element)
+    if (predicate == null || predicate(element)) {
+      result.push(element)
+    }
   }
   return result
 }
 
-export function findChildElement(element: Element, predicate: (child: Element) => boolean): Element | undefined {
+type Selector = string | string[] | ((element: Element) => boolean)
+
+function mkPredicate(root: Element, selector: Selector): (element: Element) => boolean {
+  return typeof selector === 'function'
+    ? (element: Element) => selector(element) && element !== root
+    : Array.isArray(selector)
+    ? (element: Element) => selector.includes(element.localName) && element !== root
+    : (element: Element) => element.localName === selector && element !== root
+}
+
+export function query(root: Element, selector: Selector): Element | undefined {
+  const predicate = mkPredicate(root, selector)
+
+  function go(element: Element): Element | undefined {
+    for (let i = 0, length = element.children.length; i < length; i++) {
+      const childElement = element.children[i]
+      if (predicate(childElement)) {
+        return childElement
+      }
+      const maybeElement = go(childElement)
+      if (maybeElement != null) {
+        return maybeElement
+      }
+    }
+  }
+
+  return go(root)
+}
+
+export function queryAll(root: Element, selector: Selector, recurse: boolean = true): Element[] {
+  const predicate = mkPredicate(root, selector)
+  const results: Element[] = []
+
+  function go(element: Element) {
+    for (let i = 0, length = element.children.length; i < length; i++) {
+      const childElement = element.children[i]
+      if (predicate(childElement)) {
+        results.push(childElement)
+        if (recurse) {
+          go(childElement)
+        }
+      } else {
+        go(childElement)
+      }
+    }
+  }
+
+  go(root)
+  return results
+}
+
+export function closest(element: Element, selector: Selector): Element | undefined {
```

Comment:
Perhaps `closestElement` to match the naming of (most of) the other functions in this file?

Reply:
Parent nodes may only be elements, so it’s a bit redundant in this case.

Reply:
`Parent` postfix required especially, since this function does not match to the element itself.